everest/optimizer/everest2ropt.py

```diff
@@ -1,157 +1,183 @@
+    gradients: StrListDict = defaultdict(list)
+
+    for group in controls:
+        g_sampler = _collect_sampler(group.sampler, ropt_config)
 
-        g_magnitudes = []
         for control in group.variables:
             # Collect info about the controls
-            cname = control.name
-            c_index = control.index
-            ropt_control_name = (
-                (group.name, cname) if c_index is None else (group.name, cname, c_index)
-            )
-            names.append(ropt_control_name)
-            initial_values.append(
-                control.initial_guess
-                if control.initial_guess is not None
-                else group.initial_guess
-            )
             enabled.append(
-                control.enabled if control.enabled is not None else group.enabled
+                group.enabled if control.enabled is None else control.enabled
             )
-            control_type = (
-                control.control_type
-                if control.control_type is not None
-                else group.control_type
+            cmin = group.min if control.min is None else control.min
+            cmax = group.max if control.max is None else control.max
+            scale, offset, perturbation_type = _scale_translations(
+                control.auto_scale or group.auto_scale,
+                cmin,  # type: ignore
+                cmax,  # type: ignore
+                *(control.scaled_range or group.scaled_range or (0, 1.0)),
+                group.ropt_perturbation_type,
             )
-            control_types.append(
-                VariableType.REAL
-                if control_type is None
-                else VariableType[control_type.upper()]
+            ropt_names = (  # type: ignore
+                _variable_initial_guess_list_injection
+                if isinstance(control.initial_guess, list)
+                else _variable_initial_guess_injection
+            )(
+                name=(group.name, control.name),
+                perturbation_magnitudes=(
+                    group.perturbation_magnitude
+                    if control.perturbation_magnitude is None
+                    else control.perturbation_magnitude
+                ),
+                initial_values=(
+                    group.initial_guess
+                    if control.initial_guess is None
+                    else control.initial_guess
+                ),
+                lower_bounds=cmin,
+                upper_bounds=cmax,
+                scales=scale,
+                types=(
+                    group.ropt_control_type
+                    if control.ropt_control_type is None
+                    else control.ropt_control_type
+                ),
+                perturbation_types=perturbation_type,
+                offsets=offset,
+                index=getattr(control, "index", None),
+                variables=variables,
+                gradients=gradients,
```

Comment:
I have never seen this pyhon style before. I am not sure if this is ok or very bad. I have the feeling there is something strange with the design of these classes if you have to use this to avoid code reuse

Reply:
I agree that the design of the Pydantic classes can use some refactoring. For now, this could be refractor a bit more for readability. pulling out this section of the code to its own dedicated function `_resolve_controls`